extensions/guacamole-auth-saml/src/main/java/org/apache/guacamole/auth/saml/AuthenticationProviderService.java

```diff
@@ -0,0 +1,289 @@
+                                CredentialsInfo.USERNAME_PASSWORD);
+                }
+                catch (Exception e) {
+                    logger.warn("Exception while getting name from SAML response: {}", e.getMessage());
+                    logger.debug("Received Exception while retrieving name from SAML response.", e);
+                    throw new GuacamoleInvalidCredentialsException("Error during SAML login.",
+                                CredentialsInfo.USERNAME_PASSWORD);
+                }
+            }
+        }
+
+        // No SAML Response is present, so generate a request.
+        AuthnRequest samlReq = new AuthnRequest(samlSettings);
+        URI authUri;
+        try {
+            authUri = new URI(samlSettings.getIdpSingleSignOnServiceUrl() + "?SAMLRequest=" +
+                    Util.urlEncoder(samlReq.getEncodedAuthnRequest()));
+        }
+        catch (IOException e) {
+            logger.error("Error encoding authentication request to string: {}", e.getMessage());
+            logger.debug("Got IOException encoding authentication request.", e);
+            throw new GuacamoleInvalidCredentialsException("Error during SAML login.",
+                    CredentialsInfo.USERNAME_PASSWORD);
+        }
+        catch(URISyntaxException e) {
+            logger.error("Error generating URI for authentication redirect: {}", e.getMessage());
+            logger.debug("Got URISyntaxException generating authentication URI", e);
+            throw new GuacamoleInvalidCredentialsException("Error during SAML login.",
+                    CredentialsInfo.USERNAME_PASSWORD);
+        }
+
+        // Redirect to SAML Identity Provider (IdP)
+        throw new GuacamoleInsufficientCredentialsException("Redirecting to SAML IdP.",
+                new CredentialsInfo(Arrays.asList(new Field[] {
+                    new RedirectField("samlRedirect", authUri, new TranslatableMessage("LOGIN.INFO_SAML_REDIRECT_PENDING"))
+                }))
+        );
+
+    }
+    
+    /**
+     * Generates Map of tokens that can be substituted within Guacamole
+     * parameters given a Map containing a List of attributes from the SAML IdP. 
+     * Attributes that have multiple values will be reduced to a single value,
+     * taking the first available value and discarding the remaining values.
+     * 
+     * @param attributes
+     *     The Map containing the attributes retrieved from the SAML IdP.
+     * 
+     * @return
+     *     A Map of key and single value pairs that can be used as parameter
+     *     tokens.
+     */
+    private Map<String, String> parseTokens(Map<String,
+            List<String>> attributes) {
+        
+        Map<String, String> tokens = new HashMap<>();
+        for (Entry<String, List<String>> entry : attributes.entrySet()) {
+            
+            List<String> values = entry.getValue();
+            tokens.put(entry.getKey(), values.get(0));
```

Comment:
Is it OK that these tokens will have identical names to their attributes within SAML? Or should the names be canonicalized, prefixed, etc. with http://guacamole.apache.org/doc/guacamole-ext/org/apache/guacamole/token/TokenName.html#canonicalize-java.lang.String-java.lang.String-?

Reply:
They should definitely be canonicalized with a standard prefix.